### PR TITLE
Add quick actions for contacts

### LINKED ISSUE #3365: Add quick actions for each contact
### Is your feature request related to a problem? Please describe.

Currently the information shown for each contact is just the info in the vCard and nothing else. With email and phone properties, it would be nice to directly email a contact with just one click, for example. We can show some "Quick actions" for each contact depending on the information available. 

Such quick actions also seems to be the norm for other contact management apps also, eg Apple Contacts

image

### Describe the solution you'd like

There can be some suitable actions for some fields, and if those fields are filled for a contact then the actions can be shown. For example,

Phone: Call
Email: Mail
If the contact can be linked to Talk through the system address book: Message via Talk
and more

It can be shown in the view mode for a contact, on desktop it can be on the right of the properties and on mobile it can shown below the basic info (name, role, org, photo)

image

image

### Describe alternatives you've considered

n/a

### Additional context

n/a

### PR DESCRIPTION
**Pull request by GretaD**
Screenshot from 2023-06-27 16-04-40
Screenshot from 2023-06-27 16-04-28

fixes #3365

### CODE REVIEW

File: src/components/ContactDetails.vue [hunk 1]
```diff
@@ -570,7 +624,19 @@ export default {
...
 
+		},
+		hasPhoneNumber() {
+			return this.localContact.properties.find(property => property.name === 'tel')
+		},
+		hasEmailAddress() {
+			return this.localContact.properties.find(property => property.name === 'email')
+		},
+		emailAddress() {
+			return this.localContact.properties.find(property => property.name === 'email').getFirstValue()
```
Comment (GretaD) [inline src/components/ContactDetails.vue:638]: i will refactor and find better names

File: src/components/ContactDetails.vue [hunk 2]
```diff
@@ -570,7 +625,22 @@ export default {
...
+			return this.localContact.properties.find(property => property.name === 'tel')
+		},
+		emailAddressProperty() {
+			return this.localContact.properties.find(property => property.name === 'email')
+		},
+		emailAddress() {
+			return this.emailAddressProperty.getFirstValue()
+		},
+		phoneNumber() {
+			return this.phoneNumberProperty.getFirstValue()
```
Comment (GretaD) [inline src/components/ContactDetails.vue:642]: @kesselb we had the solution right in front our eyes. getFirstValue was what we needed and not localvalue and some property value. We focused too much on the specific properties :)

File: src/components/ContactDetails.vue [hunk 3]
```diff
@@ -570,7 +624,35 @@ export default {
 
 			return ''
 		},
+		profilePageLink() {
+			// just for testing this is not fully correct
+			if (this.profileEnabled) {
+				return generateUrl('/u/{userId}', { userId: getCurrentUser().uid })
+			}
+			return null
+		},
```
Comment (ChristophWurst) [inline src/components/ContactDetails.vue:633]: analyze the result of this computed property when viewing someone else's contact

File: src/components/ContactDetails.vue [hunk 4]
```diff
@@ -570,7 +624,35 @@ export default {
...
+		emailAddressProperty() {
```
Comment (ChristophWurst) [inline src/components/ContactDetails.vue:634]:
```suggestion
		emailAddressProperties() {
```

it's none, one or more

File: lib/Controller/PageController.php [hunk 5]
```diff
@@ -108,6 +108,10 @@ public function index(): TemplateResponse {
 		$isCircleVersionCompatible = $this->compareVersion->isCompatible($circleVersion ? $circleVersion : '0.0.0', 22);
 		// Check whether group sharing is enabled or not
 		$isGroupSharingEnabled = $this->config->getAppValue('core', 'shareapi_allow_group_sharing', 'yes') === 'yes';
+		$talkVersion = $this->appManager->getAppVersion('spreed');
+		$isTalkEnabled = $this->appManager->isEnabledForUser('spreed') === true;
+
+		$isTalkVersionCompatible = $this->compareVersion->isCompatible($talkVersion ? $talkVersion : '0.0.0', 15);
```
[the remark this answers is not in the document]
Comment (GretaD) [reply lib/Controller/PageController.php:114]: no its not arbitrary, i quickly checked releases and i didnt see backports on 14th i think, but i will ask Marco or Joas for a clear answer

File: src/components/ContactDetails.vue (hunk 3 above)
Comment (GretaD) [reply src/components/ContactDetails.vue:633]: yes, this is not correct, i already tested it and i always get my uid for every user i create from my account, thats why i left the comment //testing

File: lib/Controller/PageController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply lib/Controller/PageController.php:114]: the feature was added with https://github.com/nextcloud/spreed/pull/300, so v2

**Comment by ChristophWurst** [comment]
- [ ] BUG: `[Vue warn]: Error in render: "TypeError: this.phoneNumberProperty.getFirstValue is not a function"` for contacts with no phone number

**Comment by ChristophWurst** [comment]
- [ ] BUG?: Quick actions are not aligned with the rest of the page

image

**Comment by GretaD** [comment]
> * [ ]  BUG: `[Vue warn]: Error in render: "TypeError: this.phoneNumberProperty.getFirstValue is not a function"` for contacts with no phone number

i cannot see this in my logs. I see that the icons dont have background color, are you sure you have the latest changes?

File: src/components/ContactDetails.vue [hunk 6]
```diff
@@ -570,7 +623,31 @@ export default {
...
+			return ''
+		},
+		emailAddressProperties() {
+			return this.localContact.properties.find(property => property.name === 'email')
+		},
+		emailAddress() {
+			return this.emailAddressProperties.getFirstValue()
+		},
+		phoneNumber() {
+			return this.phoneNumberProperties.getFirstValue()
```
Comment (ChristophWurst) [inline src/components/ContactDetails.vue:637]: this is where https://github.com/nextcloud/contacts/pull/3401#issuecomment-1612876117 happens. `this.phoneNumberProperties` returns `undefined` if there is no property in `this.localContact.properties` with name===tel

Comment (ChristophWurst) [inline src/components/ContactDetails.vue:637]:
```suggestion
			return this.phoneNumberProperties?.getFirstValue()
```

**Comment by JohannesGGE** [comment]
I added a fixup with the access logic to the profileURL

**Comment by GretaD** [comment]
> I added a fixup with the access logic to the profileURL

thank you Johannes, i will have a look asap

File: src/components/ContactDetails.vue [hunk 7]
```diff
@@ -221,68 +219,104 @@
...
-						:is-read-only="isReadOnly" />
+			<div v-else-if="!loadingData" class="contact-details-wrapper">
+				<div v-if="!editMode" class="quick-actions">
+					<Actions v-if="emailAddressProperties">
+						<template #icon>
+							<IconMail :size="20" />
+						</template>
+						<ActionLink v-for="emailAddress in emailAddressList"
+							:key="emailAddress"
+							:href="'email:' + emailAddress">
```
Comment (st3iny) [inline src/components/ContactDetails.vue:231]:
```suggestion
							:href="'mailto:' + emailAddress">
```

File: src/models/rfcProps.js [hunk 8]
```diff
@@ -290,6 +290,7 @@ const properties = {
 			{ id: 'YOUTUBE', name: 'YouTube', placeholder: 'https://youtube.com/…' },
 			{ id: 'MASTODON', name: 'Mastodon', placeholder: 'https://mastodon.social/…' },
 			{ id: 'DIASPORA', name: 'Diaspora', placeholder: 'https://joindiaspora.com/…' },
+			{ id: 'NEXTCLOUD', name: 'Nextcloud', placeholder: 'https://nextcloud.example.com/…' },
```
Comment (st3iny) [inline src/models/rfcProps.js:293]: For me it wasn't obvious that the profile link should be put here. Maybe add a hint to the placeholder, e.g. `Link to profile page (https://nextcloud.example.com/…)`.

**Review by st3iny — COMMENTED** [review]
I tested it briefly. Design still needs some work as the buttons are aligned incorrectly.

File: src/components/ContactDetails.vue [hunk 9]
```diff
@@ -221,68 +219,104 @@
...
+							:href="'tel:' + phoneNumber">
+							<template #icon>
+								<IconCall :size="20" />
+							</template>
+							{{ phoneNumber }}
+						</ActionLink>
+					</Actions>
+					<NcButton v-if="isTalkEnabled"
+						class="icon-talk"
+						:href="callUrl" />
```
[the remark this answers is not in the document]
Comment (kesselb) [reply src/components/ContactDetails.vue:253]: Good idea.

File: src/models/rfcProps.js (hunk 8 above)
Comment (GretaD) [reply src/models/rfcProps.js:293]: thank you both for reviewing. working on your suggestions

**Review by st3iny — APPROVED** [review]
Tested and works. The design needs some attention in a followup.

**Pull request merged**